[devgony] step2 bowling scoreboard drawing

devgony opened the pull request:
step1의 반영사항과 step2를 작성하였씁니다.
이번에도 잘 부탁드립니다! 🙇‍♂️

## Step1
### Requested changes phase2

- [x] Answers.deleteHistory 에서 DeleteHistories 받지 말고 로컬 변수로 구현
- [x] Questions.java 제거
- [x] Question.createDeleteHistory => DeleteHistory 로 이동
- [x] DeleteHistoryService.saveAll 에서 getter 로 바로 저장 하도록 수정

## Step2 - bowling scoreboard drawing

### Requirements

- 최종 목표는 볼링 점수를 계산하는 프로그램을 구현한다. 1단계 목표는 점수 계산을 제외한 볼링 게임 점수판을 구현하는 것이다.
- 각 프레임이 스트라이크이면 "X", 스페어이면 "9 | /", 미스이면 "8 | 1", 과 같이 출력하도록 구현한다.
- 스트라이크(strike) : 프레임의 첫번째 투구에서 모든 핀(10개)을 쓰러트린 상태
- 스페어(spare) : 프레임의 두번재 투구에서 모든 핀(10개)을 쓰러트린 상태
- 미스(miss) : 프레임의 두번재 투구에서도 모든 핀이 쓰러지지 않은 상태
- 거터(gutter) : 핀을 하나도 쓰러트리지 못한 상태. 거터는 "-"로 표시
- 10 프레임은 스트라이크이거나 스페어이면 한 번을 더 투구할 수 있다.

### Todos

- [x] scan Player name
    - [x] name should be 3 chars
- [x] print initial score board
- [x] rename Score -> Player, add Score enum, Player.scores = List<Score>
- [x] Player.pitch => random hit
- [x] evaluate hit with ScoreType enum, Score.payload to show current score.
- [x] Player.plays => prevScore exists, Score.play once, else twice
- [x] renew score, print score
- [x] if second of strike, iterate to next frame
    - [x] else, try same frame one more
- [x] test: add ScoreTypeTest
- [x] fix: play should hit only remainingPin
- [x] Should save, print Second score correctly.
    - [x] save only prevScore
    - [x] add gutter testcase
    - [x] save score with Hit(first, second)
- [x] add testcase for Score, Hit

### Result

```
put your nick name(3 characters): 
GON
| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |
| GON  |      |      |      |      |      |      |      |      |      |      |  
Frame 1
| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |
| GON  |  8   |      |      |      |      |      |      |      |      |      |  
Frame 1
| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |
| GON  |  8|1 |      |      |      |      |      |      |      |      |      |  
Frame 2
| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |
| GON  |  8|1 |  9   |      |      |      |      |      |      |      |      |  
Frame 2
| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |
| GON  |  8|1 |  9|/ |      |      |      |      |      |      |      |      |  
Frame 3
| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |
| GON  |  8|1 |  9|/ |  3   |      |      |      |      |      |      |      |  
Frame 3
| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |
| GON  |  8|1 |  9|/ |  3|/ |      |      |      |      |      |      |      |  
Frame 4
| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |
| GON  |  8|1 |  9|/ |  3|/ |  1   |      |      |      |      |      |      |  
Frame 4
| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |
| GON  |  8|1 |  9|/ |  3|/ |  1|3 |      |      |      |      |      |      |  
Frame 5
| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |
| GON  |  8|1 |  9|/ |  3|/ |  1|3 |  1   |      |      |      |      |      |  
Frame 5
| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |
| GON  |  8|1 |  9|/ |  3|/ |  1|3 |  1|5 |      |      |      |      |      |  
Frame 6
| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |
| GON  |  8|1 |  9|/ |  3|/ |  1|3 |  1|5 |  X   |      |      |      |      |  
Frame 7
| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |
| GON  |  8|1 |  9|/ |  3|/ |  1|3 |  1|5 |  X   |  2   |      |      |      |  
Frame 7
| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |
| GON  |  8|1 |  9|/ |  3|/ |  1|3 |  1|5 |  X   |  2|5 |      |      |      |  
Frame 8
| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |
| GON  |  8|1 |  9|/ |  3|/ |  1|3 |  1|5 |  X   |  2|5 |  2   |      |      |  
Frame 8
| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |
| GON  |  8|1 |  9|/ |  3|/ |  1|3 |  1|5 |  X   |  2|5 |  2|3 |      |      |  
Frame 9
| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |
| GON  |  8|1 |  9|/ |  3|/ |  1|3 |  1|5 |  X   |  2|5 |  2|3 |  1   |      |  
Frame 9
| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |
| GON  |  8|1 |  9|/ |  3|/ |  1|3 |  1|5 |  X   |  2|5 |  2|3 |  1|/ |      |  
Frame 10
| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |
| GON  |  8|1 |  9|/ |  3|/ |  1|3 |  1|5 |  X   |  2|5 |  2|3 |  1|/ |  7   |  
Frame 10
| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |
| GON  |  8|1 |  9|/ |  3|/ |  1|3 |  1|5 |  X   |  2|5 |  2|3 |  1|/ |  7|2 |  
```

devgony committed:
step1: refactor: Answers.deleteHistory 에서 DeleteHistories 받지 말고 로컬 변수로 구현

devgony committed:
step1: refactor: Questions.java 제거

devgony committed:
step1: refactor: Question.createDeleteHistory => DeleteHistory 로 이동

devgony committed:
step1: refactor: DeleteHistoryService.saveAll 에서 getter 로 바로 저장 하도록 수정

devgony committed:
feat: Input.scan

devgony committed:
feat: print initial score board

devgony committed:
refactor: rename Score -> Player, add ScoreType enum, Player.scores = List<Score>

devgony committed:
feat: Player.pitch => random hit, evaluate hit with ScoreType enum, Score.payload to show current score

devgony committed:
feat: Player.plays => prevScore exists, Score.play once, else twice

devgony committed:
feat: if second of strike, iterate to next frame

devgony committed:
test: add ScoreTypeTest

devgony committed:
fix: play should hit only remainingPin

devgony committed:
fix: save only prevScore, add gutter testcase

devgony committed:
fix: save score with Hit(first, second)

devgony committed:
test: add testcase for Score, Hit

devgony committed:
feat: add PlayerName,name should be 3 chars

devgony committed:
docs: add result

devgony commented on src/main/java/bowling/domain/Player.java:
```diff
@@ -0,0 +1,47 @@
+
+public class Player {
+    private final PlayerName playerName;
+    private final List<Optional<Score>> scores;
```
player 가 처음 등록 되었을때, 초기 값을 별도로 enum에 설정해두는 것을 생각했다가
Optional.empty() 로 두는 것으로 변경하였습니다. 이렇게 해도 괜찮을까요?

devgony commented on src/main/java/bowling/domain/Player.java:
```diff
@@ -0,0 +1,47 @@
+
+public class Player {
+    private final PlayerName playerName;
+    private final List<Optional<Score>> scores;
```
player 가 처음 등록 되었을때, 초기 값을 별도로 enum에 설정해두는 것을 생각했다가
Optional.empty() 로 두는 것으로 변경하였습니다. 이렇게 해도 괜찮을까요?

bingbingpa replied:
Optional 을 쓴 특별한 이유가 있을까요? 
제 생각에는 그냥 `List<Score> scores` 같은 형태로 써도 무방하지 않나하는 생각이 들어서요.

devgony commented on src/main/java/bowling/domain/Player.java:
```diff
@@ -0,0 +1,47 @@
+    public Player(String name) {
+        this.playerName = new PlayerName(name);
+        Optional<Score> empty = Optional.empty();
+        this.scores = IntStream.rangeClosed(1, 10).mapToObj(v -> empty).collect(Collectors.toList());
```
초기에 empty를 할당하는 더 나은방법이 있을지 문의 드립니다.

bingbingpa replied:
이부분도 마찬가지로 Optional 을 사용한 특별한 이유가 있다면 말씀해 주세요. 
제 생각에는 Score 변수로 두고 플레이를 할때마다 Score 인스턴스 스스로 점수를 받아서 더하면 되지 않을까 생각이 들어요.

devgony commented on src/main/java/bowling/domain/Player.java:
```diff
@@ -0,0 +1,47 @@
+            Output.print("| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |");
+            Output.print(this.payload());
+            if (newScore.done()) {
+                iterator.next();
```
스트라이크이거나, 두번 쳤을 경우에만 다음 frame으로 넘어가게 하고싶어 iterator를 사용 하였습니다. play를 재귀로 구성할까 했는데 play로직과 save 로직을 분리하기 위해서는 이 방법밖에 떠오르지 않았습니다. 혹시 더 좋은 방법이 있을까요?

bingbingpa replied:
이번 단계 미션에 있는 힌트를 잘 활용해보는건 어떨까요? 
지금은 플레이어가 스코어를 모두 관리하려다 보니 지나치게 복잡해지는 느낌이 드는거 같아요. 
~~~markDown
- 객체 단위를 가장 작은 단위까지 극단적으로 분리하는 시도를 해본다.
- 1 ~ 9 프레임을 NormalFrame, 10 프레임을 FinalFrame과 같은 구조로 구현한 후 Frame을 추가해 중복을 제거해 본다.
- 다음 Frame을 현재 Frame 외부에서 생성하기 보다 현재 Frame에서 다음 Frame을 생성하는 방식으로 구현해 보고, 어느 구현이 더 좋은지 검토해 본다.
~~~

devgony commented on src/main/java/bowling/domain/Player.java:
```diff
@@ -0,0 +1,47 @@
+                .map(v -> Score.payload(v))
+                .reduce("", (acc, cur) -> acc + cur + "|  ");
+        return payload;
+    }
```
점수판 출력하는 부분을 나름 깔끔하게 정리하고싶었으나.. 쉽지 않았습니다. 이렇게 literal로 공백을 넣어서 처리해도 괜찮은지 궁금합니다.

bingbingpa replied:
`System.out.format`을 활용하고,  적절하게 `|   ` 같은 부분들을 상수화 해보는건 어떨까요?

devgony reviewed (commented):
몇 가지 질문을 남겨드립니다.
잘 부탁 드립니다. 🙇‍♂️

devgony committed:
style: remove comments

bingbingpa commented on README.md:
```diff
@@ -20,13 +21,119 @@
+- 거터(gutter) : 핀을 하나도 쓰러트리지 못한 상태. 거터는 "-"로 표시
+- 10 프레임은 스트라이크이거나 스페어이면 한 번을 더 투구할 수 있다.
+
+### Todos
```
요구사항 정리 깔끔하게 잘해주셨네요! 💯

bingbingpa commented on src/main/java/bowling/domain/Hit.java:
```diff
@@ -0,0 +1,51 @@
+
+public class Hit {
+    private final int first;
+    private final Optional<Integer> second;
```
전반적으로 `Optional` 을 많이 사용한거 같은데요. 
Optional 을 파라미터로 사용하면 파라미터로 null 이 들어온 경우에는 
ipPresent 함수를 호출할 때 NPE 가 발생하기 때문에 파라미터로는 사용하지 않는게 좋다고 생각해요.
이와 관련해서 한번 읽어보시면 좋을거 같아서 링크 하나 공유 드릴게요. 
[링크](https://homoefficio.github.io/2019/10/03/Java-Optional-%EB%B0%94%EB%A5%B4%EA%B2%8C-%EC%93%B0%EA%B8%B0/)

devgony replied:
Optional 을 사용한 이유는 최대한 final을 사용하고 싶었는데 생성자에서 무언가를 할당 해줘야 하다 보니 -1과 같은 의미없는 숫자가 아니라 null 을 할당해서 정말 '아직 없음' 을 표현하고 싶었습니다. 하지만 남겨주신 Optional에 관한 자료를 보니 너무 과했었나 하는 생각도 듭니다 ㅠ

bingbingpa commented on src/main/java/bowling/domain/Hit.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    int remainingPin() {
+        return 10 - this.first;
```
전체적으로 매직넘버가 많이 보이는데요.
이러한 부분들을 상수화하면 가독성이 더 좋아질거 같아요.

bingbingpa commented on src/main/java/bowling/domain/Player.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    public static int pitch(int max) {
+        return new Random().nextInt(max + 1);
```
이 메소드가 호출될 때마다 `Random` 인스턴스가 생성되고 있어요.

bingbingpa commented on src/main/java/bowling/domain/PlayerName.java:
```diff
@@ -0,0 +1,17 @@
+    private String playerName;
+
+    public PlayerName(String playerName) {
+        if (playerName.length() != 3) {
+            throw new IllegalArgumentException("Player name should be 3 letter but, : " + playerName);
+        }
+            this.playerName = playerName;
```
여기도 매직 넘버는 상수화 해주는게 좋을거 같아요. 
그리고 this.playerName 에 들여쓰기도 맞춰주면 좋을거 같습니다.

bingbingpa commented on src/main/java/bowling/domain/Score.java:
```diff
@@ -0,0 +1,58 @@
+        if (score.isEmpty()) {
+            return format(" ");
+        }
+        if (score.get().scoreType == ScoreType.STRIKE) {
+            return format("X");
+        }
```
enum 을 사용하셨으니 이러한 부부은 enum 을 활용할 수도 있지 않을까 싶습니다. 
아니면 각각의투구 타입의 객체를 만들어 볼 수도 있지 않을까요?

bingbingpa commented on src/main/java/bowling/view/Input.java:
```diff
@@ -0,0 +1,11 @@
+import java.util.Scanner;
+
+public class Input {
+    static Scanner scanner = new Scanner(System.in);
```
이부분은 private 로 써주는게 좋을거 같아요. 
그리고 인스턴스 변수를 가지지 않는 유틸성 클래스라면 
인스턴스를 생성할 필요가 없기 때문에 기본 생성자를 private 로 써서 
불필요한 인스턴스 생성을 막아주는게 좋을거 같아요.

bingbingpa commented on src/main/java/bowling/domain/Score.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+
+    public static String payload(Optional<Score> score) {
```
전체적으로 `payload` 라는 메소드명을 많이 사용하셨는데요.
조금더 의미있는 메소드명을 지어보면 어떨까요?

bingbingpa commented on src/main/java/qna/domain/DeleteHistories.java:
```diff
@@ -3,23 +3,26 @@
 
-    public void add(DeleteHistory deleteHistory) {
+    public void add(Optional<Integer> idx, DeleteHistory deleteHistory) {
+        if (idx.isPresent()) {
```
1단계 피드백 잘 반영해주셨는데요. 
이부분에서 Optional 을 사용하신 특별한 이유가 있을까요? 
파라미터로 null 이 들어오면 isPresent 를 호출할 때 NPE 가 발생할거 같아요.

bingbingpa commented on src/test/java/bowling/domain/HitTest.java:
```diff
@@ -0,0 +1,49 @@
+public class HitTest {
+    Hit hit = new Hit(1);
+
+    @Test
```
테스트에 `@DisplayName` 으로 어떠한 동작을 하는 테스트인지 명확하게 
써주면 조금더 명확하지 않을까 생각합니다.

bingbingpa commented on src/main/java/bowling/domain/Player.java:
```diff
@@ -0,0 +1,47 @@
+            Score newScore = Score.play(score);
+            this.scores.set(i, Optional.of(newScore));
+            System.out.println("Frame " + (i + 1));
+            Output.print("| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |");
```
이부분은 main 에 있는 부분이랑 같은 내용 같은데요. 상수로 만들어서 같이 쓸 수 있지 않을까요? 
그리고  `Player` 가 `Output` 에 의존해서 출력을 여기에서 하는게 맞을지 한번 고민해 보면 좋을거 같아요.

bingbingpa commented on src/main/java/bowling/controller/BowlingApp.java:
```diff
@@ -0,0 +1,18 @@
+
+public class BowlingApp {
+    public static void main(String[] args) {
+        String name = Input.scan("put your nick name(3 characters): ");
```
메인에서 입력과 출력까지 모두 담당을 하고 있는거 같은데요. 
입/출력 부분은 각각 `Input`, `Output` class 가 담당하도록 해보는건 어떨까요?

bingbingpa reviewed (changes requested):
1단계 미션 피드백 반영 남겨드린거 잘 반영해주셨네요! 👍
이번 단계부터는 어려웠을텐데 빠르게 잘 구현해주셨는데요.
질문 주신 사항과 몇가지 피드백 남겨드렸는데
한번 확인해 보시고 다시 한번 리뷰 요청 부탁드립니다. 🙇
질문이나 궁금한 사항은 언제든 편하게 메시지 남겨주세요.

devgony committed:
refactor: Optional 사용 최소화

devgony committed:
refactor: 매직넘버 const 로 변경

devgony committed:
refactor: Player.pitch 내 Random 인스턴스 static으로 변경

devgony committed:
style: PlayerName indent 수정

devgony committed:
refactor: 인스턴스화 하지 않는 Input등의 유틸은 생성자를 private으로 써서 불필요한 인스턴스 생성을 막는다

devgony committed:
refactor: payload -> scoreBoard 이름 변경

devgony committed:
step1: refactor: DeleteHistories 에 Optional 제거

devgony committed:
test: 각 테스트별 @DisplayName 추가

devgony committed:
refactor: 점수판 출력에 대한 중복 부분 const화
	- 점수판 헤더
	- print 를 Output 으로 보내기
	- '|' 와 같은 foramt
	- Player가 Output에 의존하지 않도록 출력

devgony committed:
refactor: 입, 출력은 Input, Output으로 분리

devgony committed:
refactor: 플레이어가 스코어를 관리하던 역할을 분리 => Scores 일급 콜렉션으로 분리

devgony committed:
test: add test for Scores, PlayFrames, PlayFrame

devgony commented on src/main/java/bowling/controller/BowlingApp.java:
```diff
@@ -1,17 +1,20 @@
+        Output.print(HEADER_STR);
+        Output.print(player.getPlayFrame());
+        PlayFrames playFrames = player.plays();
+        Output.printFrames(playFrames);
```
main에서 Input 과 Output의 static 메서드 들을 호출만 하는것으로 뒀는데 이것 마저도 다시한번 감싸서 다른쪽에서 수행하게 해야할까요?

bingbingpa replied:
이정도도 괜찮은거 같아요.

devgony commented on src/main/java/bowling/domain/Hit.java:
```diff
@@ -1,49 +1,53 @@
     public Hit(int first) {
         this.first = first;
-        this.second = Optional.empty();
+        this.second = NULL;
```
null 을 사용하지 않으면서 final로 둘수가 없어서 초기값을 -1라는 const로 설정했는데 괜찮을까요?

bingbingpa replied:
final 로 사용고자 한다면 지금 방법도 나쁘지 않은거 같아요. 
다만 `NULL` 이라는 상수명인데 실제 값은 -1 이라서 조금 혼동이 있지 않을까 싶어요.

devgony commented on src/main/java/bowling/domain/PlayFrame.java:
```diff
@@ -0,0 +1,19 @@
+        return this.value;
+    }
+
+    public int number() {
```
Outout.printFrame 에서 값을 출력하기위해 getter를 만들었습니다. 단순 조회 후 출력이니 이정도는 괜찮을까요? 그렇지 않으면 3단계로 걸쳐서 함수 인자를 전달해야 하겠더라구요..

bingbingpa replied:
괜찮은거 같아요.

devgony commented on src/main/java/bowling/domain/Player.java:
```diff
@@ -1,47 +1,28 @@
         this.playerName = new PlayerName(name);
-        Optional<Score> empty = Optional.empty();
-        this.scores = IntStream.rangeClosed(1, 10).mapToObj(v -> empty).collect(Collectors.toList());
+        this.scores = new Scores(IntStream.rangeClosed(1, 10).mapToObj(v -> new Score()).collect(Collectors.toList()));
```
Optional 을 제거하였다보니 빈 스코어 라는 것을 표시하기 위해 new Score() {Hit(-1,-1), ScoreType.NULL} 로 구성하였습니다.

bingbingpa replied:
이부분은 `Scores`에 팩토리 메소드로 있어도 괜찮지 않을까요?

devgony commented on src/main/java/bowling/util/Const.java:
```diff
@@ -0,0 +1,15 @@
+    public final static String HEADER_STR = "| NAME |  01  |  02  |  03  |  04  |  05  |  06  |  07  |  08  |  09  |  10  |";
+    public final static String BEGIN_STR = "| ";
+    public final static String DELIMITER_STR = "  |  ";
+    public static final String END_STR = "|  ";
```
상수를 이렇게 하나의 클래스에 몰아서 사용해도 괜찮을까요?

bingbingpa replied:
하나의 클래스에 몰아서 사용을 하면 그 상수의 의미를 파악하기가 조금 힘들다고 생각해요.
그리고 여러명이 협업을 하는 경우 이 상수 클래스에 대한 수정이 많이 발생을 할꺼고
그렇다보면 충돌도 많이 발생할 거 같아요. 
상수는 각각의 필요한 클래스에 위치하는게 좋다고 생각해요.

devgony replied:
음.. 이부분은 그렇게 하고싶지만 하나의 클래스에 2개 변수 제한때문에 쉽사리 옮기기가 어렵네요 ㅠㅠ 좋은 방법이 있을까요?

bingbingpa replied:
이번 미션에서 인스턴스 변수의 수를 제한하는 이유는 하나의 클래스가 너무 많은 책임을 가지지 않도록
제한하고 있는거라고 생각하는데요. 그렇기 때문에 상수 선언은 상관 없을거 같아요.

devgony commented on src/main/java/bowling/domain/Player.java:
```diff
@@ -1,47 +1,28 @@
-    public static int pitch(int max) {
-        return new Random().nextInt(max + 1);
+    public PlayFrames plays() {
+        return this.scores.plays(this.playerName);
```
scores 일급 콜렉션을 만들면서 plays 역할을 위임했습니다. 이정도면 scores 가 직접 점수를 얻게 했다고 봐도 될까요?

devgony reviewed (commented):
빠른 피드백 감사 드립니다.
이번 변경 작업 중 가장 어려웠던 부분은 
player가 스코어를 관리하던 역할을 분리 하라고 하셨던 부분입니다.
하지만 어딘가에는 score를 저장 해야하고, 현재도 score 가 play를 하고있다고 생각해서, scores 일급 콜렉션을 만들면서 player의 대부분 메서드를 scores 로 위임 했습니다.
다만 힌트로 주신 'play 할때마다 스스로 점수를 받아서 더한다' 라는 부분이 아직 와닿지가 않아서 의도하신 대로 수정 된 것 인지 잘 모르겠습니다.. 리뷰 부탁드립니다. 🙇‍♂️

bingbingpa commented on src/main/java/bowling/domain/PlayFrames.java:
```diff
@@ -0,0 +1,20 @@
+        this.playFrames.add(playFrame);
+    }
+
+    public List<PlayFrame> playFrames() {
+        return this.playFrames;
+    }
```
컬렉션을 리턴해줄때는 불변으로 리턴해주는게 좋을거 같아요.

bingbingpa commented on src/main/java/bowling/domain/ScoreType.java:
```diff
@@ -0,0 +1,55 @@
+    public String toSymbol() {
+        return this.symbol;
+    }
+//        if (!score.isPresent()) {
+//            return format(" ");
+//        }
```
불필요한 코드는 삭제하는게 어떨까요?

bingbingpa reviewed (approved):
피드백 반영 잘해주셨네요! 👍
소소하게 남겨드린 피드백은 다음 미션 
진행하시면서 반영해보면 좋을거 같아요. 
얼마 남지 않았으니 조금만 힘내세요!
